src/test/java/com/amazon/ion/impl/bin/PresenceBitmapTest.kt

```diff
@@ -0,0 +1,330 @@
+        "One, 2, false",
+        "One, 3, false",
+        "Any, 0, true",
+        "Any, 1, true",
+        "Any, 2, true",
+        "Any, 3, false",
+        "AtMostOne, 0, true",
+        "AtMostOne, 1, true",
+        "AtMostOne, 2, false",
+        "AtMostOne, 3, false",
+        "AtLeastOne, 0, false",
+        "AtLeastOne, 1, true",
+        "AtLeastOne, 2, true",
+        "AtLeastOne, 3, false",
+    )
+    fun `validate() correctly throws exception when presence bits are invalid for signature`(cardinality: ParameterCardinality, presenceValue: Long, isValid: Boolean) {
+        val signature = listOf(Parameter("a", ParameterEncoding.uint8, cardinality))
+        with(PresenceBitmap()) {
+            initialize(signature)
+            set(0, presenceValue)
+            if (isValid) {
+                validate()
+            } else {
+                assertThrows<IonException> { validate() }
+            }
+        }
+    }
+
+    @Test
+    fun `when all parameters are tagged and exactly-one, no presence bits are needed or written`() {
+        (0..128).forEach { n -> assertExpectedPresenceBitSizes(expectedByteSize = 0, signature = List(n) { taggedExactlyOne }) }
+    }
+
+    @Test
+    fun `when all parameters are tagless and exactly-one, no presence bits are needed or written`() {
+        (0..128).forEach { n -> assertExpectedPresenceBitSizes(expectedByteSize = 0, signature = List(n) { taglessExactlyOne }) }
+    }
+
+    @Test
+    fun `when all parameters are tagged and not exactly-one, should write expected number of presence bits`() {
+        // Index of an element in this list is the number of parameters in the signature
+        listOf(0, 0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3)
+            .forEachIndexed { numParameters, expectedByteSize ->
+                assertExpectedPresenceBitSizes(expectedByteSize, signature = List(numParameters) { taggedZeroToMany })
+            }
+    }
+
+    @Test
+    fun `when all parameters are tagless and not exactly-one, should write expected number of presence bits`() {
+        // Index of an element in this list is the number of parameters in the signature
+        listOf(0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3)
+            .forEachIndexed { numParameters, expectedByteSize ->
+                assertExpectedPresenceBitSizes(expectedByteSize, signature = List(numParameters) { taglessZeroToMany })
+            }
+    }
+
+    private fun assertExpectedPresenceBitSizes(expectedByteSize: Int, signature: List<Parameter>) {
+        val result = writePresenceBits { pb ->
+            pb.initialize(signature)
+            assertEquals(expectedByteSize, pb.byteSize)
+            assertEquals(signature.size, pb.totalParameterCount)
+        }
+        assertEquals(expectedByteSize, result.size)
+    }
+
+    @Test
+    fun `read 4 parameters`() {
+        val signature = listOf(taggedZeroToMany, taggedZeroOrOne, taggedExactlyOne, taggedZeroToMany)
+
+        val bytes = bitStringToByteArray("00100010")
```

Comment:
I was confused by this at first because I was trying to map the four slots here to the four elements of the signature, but only three of the elements of the signature correspond to presence bits, while the exactly-one is implicit. The comment below implies that, but wasn't clear enough for me to get it the first time. Consider adding a comment before this line to explain that the bits are read in pairs from right-to-left and are only present for non-`!` signature elements. It would also help if it were `bitStringToByteArray("100010")`, assuming this will be automatically zero-padded anyway.